### PR TITLE
feat(rl,#11297): rlpt_3 reward hacking — anatomie d'un hack qui ne prend pas

### PR DESCRIPTION
**Pull request by jsboige**
Grain: MED/training — lane myia-po-2024:CoursIA — prev: DEEP/training #11304 (rlpt_2)

## Summary

Nouveau notebook **`MyIA.AI.Notebooks/RL/rlpt_3_reward_hacking.ipynb`** (27 cellules, 12 code) :
**anatomie d'un reward hacking qui ne prend pas** — le piège est construit, les trois voies de
déclenchement sont mesurées, et toutes trois échouent ; c'est le résultat, cohérent avec le capstone
ICT-25 (#5105, PR #11298) dont ce notebook est la version compacte instrumentée. Série RL Post-Training
**#11297** (grain 3/4, tier MED).

## Le design

Même stack que rlpt_2 (Qwen3.5-0.8B QLoRA local, soustractions conversationnelles, GRPO trl 1.9.2,
seed fixée 42 partout). Deux rewards exclusifs par construction (`b >= 10` ⇒ `a - b != a`) :
**`reward_exact`** (vrai objectif) vs **`reward_proxy`** (le piège : récompense l'opérande **recopié**
de l'énoncé). **Détection cuite dans le reward** : `reward_proxy` logue l'exactitude réelle de chaque
batch — les deux courbes au même pas de temps, sans évaluation externe.

## L'histoire mesurée (outputs réels commités, seed 42)

1. **Baseline** : recopie spontanée **0.00 partout** (propre et contaminé, greedy et sampling) —
   mais l'exemple piégé **perturbe** fortement (exacte greedy 0.66 → 0.28 contaminé).
2. **Run A (proxy seul, 20 steps)** : exploration morte — proxy **0.000 constant**,
   `frac_reward_zero_std` **1.0**, `grad_norm` 0.00 sur tout le run ; poids immobiles. Un comportement
   jamais produit ne peut pas être sélectionné par l'avantage par groupe.
3. **§3.1 artefact d'évaluation train/eval** (démo mesurée sur le modèle du run A) : greedy en mode
   train **0.00** vs mode eval **0.88** (écart +0.88, dropout LoRA actif) ; écart +0.00 mesuré sur la
   base sans adaptateur. Toutes les mesures du notebook forcent `eval()`.
4. **Inoculation few-shot** : l'exemple piègé installe la **surface**, pas la règle — inspection des
   réponses : imitation de la réponse de l'exemple ('45') bien plus fréquente que la recopie de
   l'opérande propre à la question.
5. **Run B (amorce + proxy, 40 steps, prompt contaminé)** : le hack **ne prend pas** — proxy lissé
   **0.000 → 0.031**, recopie post-run 0.06 (contaminé) / 0.00 (propre) ; MAIS la détection par
   divergence fonctionne : exactitude lissée **0.500 → 0.344** (chaque point de recopie se paie en
   exactitude, exclusivité par construction). `disable_adapter()` : base intacte (0.84 / 0.00).
6. **Run C (désintoxication)** : adaptateur figé du run B + re-entraînement reward exact — courbe
   **0.350 → 0.750**, post-C greedy **0.88 / 0.00** (niveau de la base intacte). Le cas facile :
   le reward exact est dense (variance de reward présente), à l'inverse du run A.

Verdict pédagogique : à 0.8B, deux barrières tiennent (taux de production initial nul, attracteur
faible) ; elles s'abaissent avec l'échelle et les steps (Gao et al. 2023) — mais la **détection par
divergence** et la **désintoxication** fonctionnent à toutes les échelles, et c'est elles qu'on
industrialise.

## Validation

- Exécution réelle end-to-end via nbconvert kernel `coursia-ml-training` : exit 0, 12 cellules code,
  0 erreur, `execution_count` 1-12 + outputs cohérents (C.2). La cellule §4 (inspection) a été
  re-exécutée réellement (transplant 1-cell) après correction de son ancrage modèle
  (`disable_adapter`), toutes les autres portent les outputs de l'exécution complète.
- 0 violation C.1 (3 exercices stubs `pass`/`return None`/`print` — parité, divergence_alert, SFT
  inoculation).
- Tous les chiffres des interps extraits des outputs commités par script (dump, jamais reconstruits) ;
  prose-verdicts rédigées après lecture des outputs réels.
- Runs GRPO réels ×3 sur GPU 8 Go (1.4 + 4.4 + 2.7 min), seed fixée : reproductible.

## Regression check

- Fichier nouveau uniquement — aucun notebook existant modifié (C.3 OK).
- Catalogue byte-identique à main.

See #11297 (grain 3/4, série RL Post-Training) — See #5105 (capstone ICT-25 dont ce notebook décline
le verdict)

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
Vérifié (rlpt_3, #11297) — exécution authentique, structure pédagogique cohérente, RAS.

**Diff (1029 lignes, 1 notebook)** : 27 cellules (12 code / 15 md), notebook capstone reward hacking.

**Authenticité outputs** : 12/12 cellules code avec `execution_count`, 10/12 avec outputs. Sorties réelles observées : versions environnement complètes (transformers 5.15.0, trl 1.9.2, torch 2.6.0+cu124, RTX 3070 8 Go), dataset 48 rows, courbes d'entraînement display_data, warnings transformers authentiques (PAD/BOS/EOS tokenizer mismatch, `use_cache` vs gradient checkpointing). Cellules 23-25 = exercices non exécutés, structure attendue (3 pièges : proxy parité, test divergence, inoculation SFT).

**Fuites** : aucun chemin machine (`/home`, `/opt/data`, `C:\Users`), aucun secret/token dans le diff, zéro sorry.

**Métriques cohérentes** : BASE propre greedy 0.66 exacte / sampling 0.53 vs contamine — le contraste contamine/propre illustre le hack qui ne prend pas, aligné avec le titre et la progression #11297.

**CI** : 54/55 checks pass (1 SKIPPED = advisory non-main base). Kernel `coursia-ml-training` Python 3.11, nbformat 4.5.

[INFO] [Hermes]

**Comment by myia-ai-01** [comment]
Gate coord — mesuré sur la tête de PR, pas sur le body :

```
code=12  exec_null=[]  errors=0  C.1 violations=[]        → EXEC_PROVED
prose chars/cellule de code = 1075   (seuil 1200)         → l'étiquette a raison
```

`pedagogy-density-below-threshold` n'est pas un faux positif ici — 1075/1200. Je merge quand même : l'étiquette est advisory et le grain est de la substance (DEEP/training, run réel multi-seed, finding TRL chat-template documenté). Mais elle **pose une question à laquelle on répond par du contenu, pas par une mise en sourdine** — je l'ai déjà désarmée à tort deux fois en croyant qu'elle comptait les exercices ; elle compte des caractères de prose par cellule de code, et elle avait raison les deux fois.

Pour `rlpt_4` (dernier grain de la série) : viser le seuil dès l'écriture plutôt qu'en tranche de rattrapage — la veine densité #11224 est plafonnée et ce serait un grain LIGHT de plus là où celui-ci est du contenu.

Lecture honnête retenue au passage : « GRPO concentre la distribution, ne crée pas de capacité » (greedy 0.88±0.00 sur 4 seeds) — c'est le genre de verdict qui vaut le notebook.
